example_queries/envoy/height.rs.ref

```diff
@@ -69,7 +69,7 @@ pub fn execute_udfs_and_check_trace_lvl_prop(http_headers: &HttpHeaders, fd: &mu
                 petgraph::Outgoing);
             let mut child_values = Vec::new();
             for child in child_iterator {
-                child_values.push(fd.trace_graph.node_weight(child).unwrap().1["height"].clone());
+                child_values.push(fd.trace_graph.node_weight(child).unwrap().1["height(a)"].clone());
```

Comment:
I am very suspicious of this.  Properties of nodes should not contain structural filter nodes.  What does it even mean if node c then has a height(a) attribute?

Reply:
This is the return label or reference to the UDF result. The query looks like this 
`MATCH (a) -[]-> (b {} )-[]->(c) RETURN height(a)`
So this is actually a UDF argument and not a structural filter. To be honest the entire design is a little weird, I am not sure what to make of it. I realized that RETURN makes the implicit assumption that the result is tied to a node or entity in the graph (so `a.property` or `a.udf_call`). So we implicitly select a node already.  In our design we somehow compute UDF results and attach them to nodes in the tree, carry them along, and then select them based on the RETURN statement.  All of this is confusing. Technically, you should only execute the RETURN when the match applies. What should be carried along is `a` and then you compute the result with that input, but that is too vague of an object. 

Maybe it is better to express this clearly so it can also be explained in simple terms.
The way these queries work currently is actually comparable to this: 
`CALL height() YIELD height MATCH a) -[]-> (b {} )-[]->(c) RETURN a.height`
where you first call height on ALL nodes without any arguments, which creates a new property for each node, then you filter, then you access that property. This is the best way I could express this in cypher terms. 

Another way to express this to to have a second selection step: 
`MATCH (a) -[]-> (b {} )-[]->(c) WITH a RETURN height(a)`
here we first filter for the pattern, then select the node that is relevant, then compute the result on it.
There is many ways to express this but we should make explicit that calling a UDF actually leads to execution on all path and that the result is actually a new property in the graph.

Reply:
I think the first of those two gives the clearest idea of what we're doing.  The only issue is, how would that work for multiple UDFs?  We should probably also show this to Anirudh or someone less down in the weeds to make sure it's still intuitive for someone who hasn't really seen it before.

Reply:
You can chain these calls. It basically describes the way scalar UDF calls work internally now. If the OpenCypher syntax becomes to unwieldy we can also change it, we just need to change the ANTLR spec and regenerate a parser. 

Yeah let's do that. What I believe is most important is a coherent query design that does not cause surprises and is constrained appropriately. So the query should reflect in some part how we internally evaluate calls and expressions. This also makes parsing much, much easier, and makes it also simpler for us to classify segments of the query into the right IR constructs, that we can use to apply optimizations.

Reply:
Okay.  We can get the changes in now so we have the udf(x) syntax down and then we can propose this syntax at the next meeting.  Are you good with that as a plan?

Reply:
Yeah sounds good. The simulator still needs to be adjusted, so that might take a little.
I also have a temporary fix for the serialization issue, I can attach it as a rider to this pull request. The long term fix to the serialization issue is using a type system for specific properties. Either hardcoded or provided by the query.